port-forwarder/fwd_tcp.go

```diff
@@ -0,0 +1,230 @@
+	return fmt.Sprintf("outbound.tcp.%s.%s", cfg.localListen, cfg.remoteConnect)
+}
+
+type ForwardConfigIncomingTcp struct {
+	ForwardConfigIncoming
+}
+
+func (cfg ForwardConfigIncomingTcp) ConfigDescriptor() string {
+	return fmt.Sprintf("inbound.tcp.%d.%s", cfg.port, cfg.forwardLocalAddress)
+}
+
+type PortForwardingOutgoingTcp struct {
+	l                     *logrus.Logger
+	tunService            *service.Service
+	cfg                   ForwardConfigOutgoingTcp
+	localListenConnection *net.TCPListener
+}
+
+func (fwd PortForwardingOutgoingTcp) Close() error {
+	fwd.localListenConnection.Close()
+	return nil
+}
+
+func (cf ForwardConfigOutgoingTcp) SetupPortForwarding(
+	tunService *service.Service,
+	l *logrus.Logger,
+) (io.Closer, error) {
+	localTcpListenAddr, err := net.ResolveTCPAddr("tcp", cf.localListen)
+	if err != nil {
+		return nil, err
+	}
+	remoteTcpAddr, err := net.ResolveTCPAddr("tcp", cf.remoteConnect)
+	if err != nil {
+		return nil, err
+	}
+	localListenPort, err := net.ListenTCP("tcp", localTcpListenAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	l.Infof("TCP port forwarding to '%v': listening on local TCP addr: '%v'",
+		remoteTcpAddr, localTcpListenAddr)
+
+	portForwarding := &PortForwardingOutgoingTcp{
+		l:                     l,
+		tunService:            tunService,
+		cfg:                   cf,
+		localListenConnection: localListenPort,
+	}
+
+	go portForwarding.acceptOnLocalListenPort()
+
+	return portForwarding, nil
+}
+
+func (pt *PortForwardingOutgoingTcp) acceptOnLocalListenPort() error {
+	for {
+		pt.l.Debug("listening on local TCP port ...")
+		connection, err := pt.localListenConnection.AcceptTCP()
+		if err != nil {
+			fmt.Println(err)
+			return err
+		}
+
+		pt.l.Debugf("accept TCP connect from local TCP port: %v", connection.RemoteAddr())
+
+		go pt.handleClientConnection(connection)
+	}
+}
+
+func (pt *PortForwardingOutgoingTcp) handleClientConnection(localConnection *net.TCPConn) {
+	err := pt.handleClientConnection_intern(localConnection)
+	if err != nil {
+		pt.l.Debugf("Closed TCP client connection %s. Err: %v",
+			localConnection.LocalAddr().String(), err)
+	}
+}
+
+func (pt *PortForwardingOutgoingTcp) handleClientConnection_intern(localConnection net.Conn) error {
+
+	remoteConnection, err := pt.tunService.DialContext(context.Background(), "tcp", pt.cfg.remoteConnect)
+	if err != nil {
+		return err
+	}
+	return handleTcpClientConnection_generic(pt.l, localConnection, remoteConnection)
+}
+
+func handleTcpClientConnection_generic(l *logrus.Logger, connA, connB net.Conn) error {
+
+	dataTransferCtx, cancel := context.WithCancel(context.Background())
+
+	dataTransferHandler := func(from, to net.Conn) error {
+
+		name := fmt.Sprintf("%s -> %s", from.LocalAddr().String(), to.LocalAddr().String())
+
+		defer from.Close()
+		defer to.Close()
+		defer cancel()
+		// give communication in opposite direction time to finish as well
+		defer time.Sleep(time.Millisecond * 300)
```

Comment:
I think you want to avoid sleeps like this one if possible

Reply:
I agree, especially in other languages like C++ where the default is synchronous execution. With golang I wasn't so sure whether its a real problem as the goroutines seems to compare to the async sleeps of other languages.

I honestly do not know it this is actually really needed at all. Because this is intented to catch the case where the receiving side is closed before the sending side could send all requested data. I do not have a testing possibility for this. Do you know if this is needed, or how to test it ideally?

Reply:
I removed it.